### ISSUE TITLE
Reverso in Darts

### ISSUE THREAD

**Issue by daidahao** [shinfxh/reverso#1]
Dear @shinfxh and Reverso team,

Congratulations on the Reverso model release and the paper! It's truly remarkable how such a tiny model achieve accuracy on par with models with 100M+ parameters.

[Darts](https://github.com/unit8co/darts) is a popular time series library and provides sklearn-like unified forecasting APIs. For the past few months, I've been driving the efforts of integrating latest foundation models into Darts, such as [Chronos-2](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.chronos2_model.html#darts.models.forecasting.chronos2_model.Chronos2Model) and [TimesFM 2.5](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.timesfm2p5_model.html#darts.models.forecasting.timesfm2p5_model.TimesFM2p5Model), with fine-tuning support on the horizon as well.

It would be great for our users to use Reverso within Darts, given Reverso's strong zero-shot accuracy and outstanding parameter efficiency. Would you be open to contributing a reference implementation in Darts? This would also help boost Reverso’s visibility and adoption by making it directly accessible to Darts users. If you’re interested, I’d be happy to assist with implementation guidance, review, and testing to the best of my ability.

In terms of technicality, foundation models in Darts use [`FoundationModel`](https://github.com/unit8co/darts/blob/bcfc0fd4095a4343b7c3f4861ffe6f47223c3ee6/darts/models/forecasting/foundation_model.py#L21) as a base class for PyTorch-native implementation and `HuggingFaceConnector` to load model weights from HuggingFace Hub. So you may use them to compose Reverso cleanly and leave it to Darts' [TorchForecastingModel](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html) to handle auto-regression, historical forecasting, fine-tuning, and more.

**A word of concern:** Looking at current Reverso codebase, I noticed dependencies on `FlashFFTConv` and `flash-linear-attention`, which could impose additional dependencies and hardware constraints (i.e., CUDA-only) and limit the potential user pool. If you are open to the idea here, would it be possible to also provide a PyTorch-native alternatives to `FlashFFTConv` and `DeltaNet`?

Let me know what you think. Thank you for open-sourcing Reverso!

Best,
Zhihao

**Comment by shinfxh** [comment]
Hi Zhihao,

Thanks for the great work on Darts! We're happy to contribute to integration into Darts.

Indeed, the hardware dependencies might be a bit restricting, but maybe we can start with the existing implementation in this repo in the meantime while I add more support?

Let me know where's the best place to start.

Best,
Xinghong

**Comment by daidahao** [comment]
Hi @shinfxh 

Great to hear you are on board! The best way to start would be looking at current TSFM implementations in Darts, before composing Reverso via `FoundationModel` base class :
- [chronos2_model.py](https://github.com/unit8co/darts/blob/45c2eec1a94521dc3ba4da72b5e615512c9a0dac/darts/models/forecasting/chronos2_model.py)
- [timesfm2p5_model.py](https://github.com/unit8co/darts/blob/45c2eec1a94521dc3ba4da72b5e615512c9a0dac/darts/models/forecasting/timesfm2p5_model.py)

I've looked deeper into `FlashFFTConv` and `flash-linear-attention` codebase and it appears that they don't offer CPU-compatible or PyTorch-native `forward()` implementation. That worries me because Darts CI/CD is CPU-only and will not be able to test Reverso for integration. We also would like to offer Reverso to as many users as possible without hefty hardware restrictions. I wonder if there is any way we could resolve this before starting a PR in Darts.

**Comment by shinfxh** [comment]
Just included the torch-native implemenation in the most recent [commit](https://github.com/shinfxh/reverso/commit/9e74d1915604558fdc9e8938e44b3af581f8be55)!

Compared to the original implementation, the torch-native version has around a 4x slowdown due to unoptimized kernels even on GPU.

There are also some slight deterioration due to floating point mismatch but it is within tolerance. 

Some replicated numbers for MASE are:

Original implementation: 0.7263
Torch-GPU: 0.7264
Torch-CPU: 0.7272

**Comment by daidahao** [comment]
@shinfxh 

Thank you for the port! I would start an issue in Darts library to align the API design between you and Darts owners and to host discussions, a standard process before starting a PR in Darts.

Once you have started a PR, I am happy to provide thorough testing and feedback. If you need any help, I am also here to provide clarification on Darts APIs, implementation and review.

P.S. For the floating point mismatch, is it due to the original using bfloat16 vs torch-native using flaot32? If so, perhaps we could use [`precision`](https://lightning.ai/docs/pytorch/stable/common/precision_intermediate.html) option in PyTorch Lightning trainer, which Darts uses to control torch model logic, to ensure precision alignment and regain some of those speed lost.

PS2. Could you confirm that both the model weights and code are released under the MIT license, so they are compatible with Apache 2.0 used by Darts?
